src/ObjectUtils/CoffFile.cpp

```diff
@@ -417,49 +417,130 @@ orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
-  // For a function, the "primary" RUNTIME_FUNCTION and the ones with UNW_FLAG_CHAININFO set (if
-  // any) form a tree rooted on the "primary" RUNTIME_FUNCTION. We merge all such RUNTIME_FUNCTIONs
-  // from the same tree into a single UnwindRange.
+  uint64_t last_primary_address{};
+  uint64_t largest_primary_address{};
+  uint64_t num_chained_not_with_largest_primary = 0;
   for (const llvm::Win64EH::RuntimeFunction& runtime_function : runtime_functions) {
     auto primary_runtime_function_or_error = GetPrimaryRuntimeFunction(&runtime_function);
     if (primary_runtime_function_or_error.has_error()) {
-      return ErrorMessage{absl::StrFormat("Unable to load unwind info ranges: %s",
-                                          primary_runtime_function_or_error.error().message())};
+      return ErrorMessage{
+          absl::StrCat(kErrorMessagePrefix, primary_runtime_function_or_error.error().message())};
     }
     const llvm::Win64EH::RuntimeFunction* primary_runtime_function =
         primary_runtime_function_or_error.value();
 
+    const uint64_t start_address = object_file_->getImageBase() + runtime_function.StartAddress;
+    const uint64_t end_address = object_file_->getImageBase() + runtime_function.EndAddress;
     const uint64_t primary_address =
         object_file_->getImageBase() + primary_runtime_function->StartAddress;
-    const uint64_t end_address = object_file_->getImageBase() + runtime_function.EndAddress;
 
     if (unwind_ranges.empty()) {
       unwind_ranges.emplace_back(UnwindRange{primary_address, end_address});
+      ORBIT_CHECK(last_primary_address == 0);
+      ORBIT_CHECK(largest_primary_address == 0);
+      last_primary_address = primary_address;
+      largest_primary_address = primary_address;
       continue;
     }
 
+    ORBIT_CHECK(last_primary_address != 0);
+    ORBIT_CHECK(largest_primary_address != 0);
+
+    UnwindRange& last_range = unwind_ranges.back();
+    const uint64_t last_end_address = last_range.end;
+
+    if (end_address < start_address) {
+      return ErrorMessage{
+          absl::StrCat(kErrorMessagePrefix, "RUNTIME_FUNCTION with negative function size.")};
+    }
+
     // https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64#struct-runtime_function
     // guarantees that RUNTIME_FUNCTIONs are sorted (by address), and we rely on that here.
-    UnwindRange& last_range = unwind_ranges.back();
-    if (primary_address == last_range.start) {
-      // Extend the size of the current function.
+    if (start_address < last_end_address) {
+      return ErrorMessage{
+          absl::StrCat(kErrorMessagePrefix, "RUNTIME_FUNCTIONs not sorted or overlapping.")};
+    }
+
+    // https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64#struct-unwind_info states
+    // that "the chained unwind info entry is the contents of a *previous* RUNTIME_FUNCTION entry",
+    // and we rely on that here.
+    if (primary_address > start_address) {
```

Comment:
Nit: Other checks use `<`. I would find it slightly more intuitive to read if this were also `start_address < primary_address`.